p4-16/spec/P4-16-spec.adoc

```diff
@@ -3577,10 +3577,18 @@ Bit-strings also support the following operations:
   `+0 <= L <= H < W+` are checked statically (where `W` is
   the length of the source bit-string). Note that both endpoints of
   the extraction are inclusive. The bounds are required to be
-  local compile-time known values so that the width of the result can be computed at compile time. Slices are also l-values, which means that P4 supports assigning to a slice: `e[H:L] = x`.
+  local compile-time known values so that the width of the result can be computed
+  at compile time. Slices are also l-values, which means that P4 supports
+  assigning to a slice: `e[H:L] = x`.
   The effect of this statement is to set bits `H` through `L` (inclusive of
   both) of `e` to the bit-pattern represented by `x`, and leaves all other bits
-  of `e` unchanged.  A slice of an unsigned integer is an unsigned integer.
+  of `e` unchanged.  A slice is always an unsigned integer.
+  A slice may also be specified as `[L+:W]` where `L` is the lowest bit of the slice
+  and `W` is the width of the slice.  In this case, only `W` must be a non-negative
+  local compile-time known numeric value.  `L` must also be a numeric value and
+  some architectures may require it to be compile-time known.  It must also be
+  non-negative and in-range for the type.  Some architectures may allow variable indexing,
+  in which case an out of range value will be equivalent to a shift of that amount.
```

Comment:
I suggest wording this as for shifts, allowing only `int` or `bit<W>`, not `int<W>` (https://p4.org/wp-content/uploads/2024/10/P4-16-spec-v1.2.5.html#sec-a-note-about-shifts). This also matches what was discussed on the last LDWG. We cannot defer to the shift for the `int<W>` index as shift does not allow `int<W>` to avoid probelems with negative values.

Reply:
Maybe add a precise definition for "non-negative" somewhere -- either a `bit<W>` value or a compile time known value that is >= 0.   I don't think we need to (or should) specifically disallow `int<W>` compile-time-known values.

Reply:
I think the formulation that is used for RHS of shifts permits that:

> [...] the right operand of shifts must be either an expression of type bit<S> or a compile-time known value that is a non-negative integer.

I suggest we use the same.

I think a restriction to operation types is something important enough to have it written in the operation description and not hidden in a definition of non-negativity.

Reply:
@ChrisDodd Any thoughts on updates to this PR to address this comment?

Reply:
I added a line here to clarify that the index must be a `bit<W>` type if it is not compile-time-known, in order to be compatible with shift.

Reply:
I see.  @vlstill Any additional comments, or does it look good to you?